docs/DynamicCasting.md

```diff
@@ -0,0 +1,532 @@
+Except as described below, casting between dissimilar types always fails -- for example, casting a struct instance to a function type or vice-versa.
+
+Where possible, each section includes machine-verifiable _invariants_ that can be used as the basis for developing a robust test suite for this functionality.
+
+## Identity Cast
+
+Casting an instance of a type to its own type will always succeed and return the original value unchanged.
+
+```
+let a: Int = 7
+a is Int // true
+a as? Int // Succeeds
+a as! Int == a // true
+```
+
+## Classes
+
+Casting among class types follows standard object-oriented programming conventions:
+
+* Class upcasts: If `C` is a subclass of `SuperC` and `c` is an instance of `C`, then `c is SuperC == true` and `(c as? SuperC) != nil`.  These "upcasts" do not change the representation in any way.  However, when `c` is accessed via a variable or expression of type `SuperC`, only methods and instance variables defined on `SuperC` are available.
+
+* Class downcasts: If `C` is a subclass of `SuperC` and `sc` is an instance of `SuperC`, then `sc is C` will be true iff `sc` is actually an instance of `C`.  When this "downcast" does succeed, it does not change the representation.
+
+* Objective-C class casting:  The rules above also apply when one of the classes in question is defined with the `@objc` attribute or inherits from an Objective-C class.
+
+* Class casts to AnyObject: Any class reference can be cast to `AnyObject` and then cast back to the original type.  See "AnyObject" below.
+
+* If a struct or enum type conforms to `_ObjectiveCBridgeable`, then classes of the associated bridging type can be cast to the struct or enum type and vice versa.  See "The `_ObjectiveCBridgeable` Protocol" below.
+
+Invariants:
+* For any two class types `C1` and `C2`: `c is C1 && c is C2` implies that `(c as! C1) is C2`
+* For any class type `C`: `c is C` implies that `(c as! AnyObject) as! C === c`
+
+## Structs and Enums
+
+You cannot cast between distinct struct or enum types.
+More formally:
+
+* If `S` and `T` are struct or enum types and `s is S == true`, then `s is T` iff `S.self == T.self`.
+
+## Tuples
+
+Casting from a tuple type T1 to a tuple type T2 will succeed iff the following hold:
+* T1 and T2 have the same number of elements
+* If an element in T1 has a label, the corresponding element in T2 has the same label
+* Each element of T1 can be individually cast to the corresponding type of T2
+
+## Functions
+
+Casting from a function type F1 to a function type F2 will succeed iff the following hold:
+* The two types have the same number of arguments
+* Corresponding arguments have identical types
+* The return types are identical
+* If F1 is a throwing function type, then F2 must be a throwing function type.  If F1 is not throwing, then F2 may be a throwing or non-throwing function type.
+
+Note that it is _not_ sufficient for argument and return types to be castable; they must actually be identical.
```

Comment:
```
class A {}
class B: A {}
class C: B {}

let x: (B) -> B = { $0 }
let y = x as! (C) -> A
```

Reply:
```
% swift
Welcome to Apple Swift version 5.3 (swiftlang-1200.0.16.9 clang-1200.0.22.5).
Type :help for assistance.
  1> class A {} 
  2. class B: A {} 
  3. class C: B {} 
  4.  
  5. let x: (B) -> B = { $0 } 
  6. let y = x as! (C) -> A
Could not cast value of type '(__lldb_expr_1.B) -> __lldb_expr_1.B' (0x7fff97c994f8) to '(__lldb_expr_1.C) -> __lldb_expr_1.A' (0x7fff97c99530).
2020-07-20 16:21:33.275169-0700 repl_swift[49292:29208408] Could not cast value of type '(__lldb_expr_1.B) -> __lldb_expr_1.B' (0x7fff97c994f8) to '(__lldb_expr_1.C) -> __lldb_expr_1.A' (0x7fff97c99530).
x: (B) -> B =
y: (C) -> A =
```

Reply:
Huh, this works in 5.1 (I still have Xcode 11 as my daily driver)
```
Apple Swift version 5.1 (swiftlang-1103.0.1 clang-1103.0.17.90)
Target: x86_64-apple-darwin19.2.0
```

Good to know!

Reply:
Yeah.  There's been a bit of churn in casting behaviors since 5.0.  One goal of this specification is to help reduce that going forward.

Reply:
Another clarification:  I tested this again as a Swift source file compiled with and without optimizations and got the same result against Swift 5.3 as I did in the REPL above.  (I just wanted to make sure this wasn't one of the places where optimized and unoptimized casts behaved differently.)